autosubmit/notifications/mail_notifier.py

```diff
@@ -53,26 +65,34 @@ def notify_status_change(self, exp_id, job_name, prev_status, status, mail_to):
                 Log.printlog('Trace:{0}\nAn error has occurred while sending a mail for the job {0}'.format(e,job_name), 6011)
 
     def _send_mail(self, mail_from, mail_to, message):
-        server = smtplib.SMTP(self.config.SMTP_SERVER,timeout=60)
+        server = smtplib.SMTP(self.config.SMTP_SERVER, timeout=60)
         server.sendmail(mail_from, mail_to, message.as_string())
         server.quit()
 
+    def _attach_files(self, message, files):
+        for f in files or []:
+            with open(f, "rb") as file:
+                part = MIMEApplication(file.read(), Name = Path(f).name)
+                part['Content-Disposition'] = 'attachment; filename="%s"' % Path(f).name
+                message.attach(part)
```

Comment:
If we use type hints in the rest of the code changed here, then we would be sure that `files` is a `List[Path]` (or as sure as we can get).

If we do not specify it's `Optional` or `Union[List[Path], None]`, then we are declaring the caller must prevent `None`s, and it eliminates the need for `in files or []` (but a good idea to be defensive, IMO).

Knowing we have `Path`s in the list, then we don't need to use a context-manager, so no need for the `with open(f, "rb") as file:`, and instead we can just read the content directly with `pathlib`'s utility functions. And also no need to re-wrap the `Path` object `f` in a new `Path`. Combined with f-strings, I think the new code can be simplified (with type hints) down to something like:

```diff
    def _attach_files(self, message, files: List[Path]) -> None:
        for f in files:
            part = MIMEApplication(f.read_text(), Name=f.name)
            part['Content-Disposition'] = f'attachment; filename="{f.name}"'
            message.attach(part)
```